Sources/bonfire/flame/Parser.hx

```diff
@@ -0,0 +1,250 @@
+package bonfire.flame;
+
+import bonfire.flame.Literal;
+import bonfire.flame.Stmt;
+import bonfire.flame.Expr;
+import bonfire.flame.Token;
+import bonfire.flame.utils.Printer.trace in err;
+import bonfire.flame.utils.Iterators.PeekableTokenIterator in PTI;
+
+class Parser {
+    static var hadError: Bool;
+
+    public static function parse(tokens: Array<Token>): Array<Stmt> {
+        var titer = new PTI(tokens);
+        hadError = false;
+        var stmts = new Array<Stmt>();
+        var stmt: Stmt = null;
+        while (!titer.eatIf(EOF)) {
+            try {
+                stmt = statement(titer);
+                if (stmt == null) {
+                    throw new ParseError(titer.peek(), 'Expected start of statement, got unexpected ${titer.peek().type}');
+                }
+            } catch(e: ParseError) {
+                err(e.message, e.token);
+                titer.index++;
+                while ((try {
+                    (stmt = statement(titer)) == null;
+                } catch (e: ParseError) {
+                    true;
+                })) {
+                    if (titer.eatIf(EOF)) {
+                        return stmts;
+                    }
+                    titer.index++;
+                }
+            }
+            stmts.push(stmt);
+        }
+        return stmts;
+    }
+
+    static inline function statement(titer: PTI): Stmt {
+        switch (titer.peek().type) {
```

Comment:
`switch` is an expression in Haxe, so instead of repeating `return` in every branch, you can just do `return switch ...`. ;)

https://code.haxe.org/category/principles/everything-is-an-expression.html